libparsec/crates/client/src/certif/shamir_recovery_setup.rs

```diff
@@ -0,0 +1,363 @@
+    author_device: DeviceID,
+    author_id: UserID,
+    share_recipients: &HashMap<UserID, u8>,
+    threshold: u8,
+    timestamp: DateTime,
+) -> Result<ShamirInternalsOutcome, CertifShamirError> {
+    // 1. Check if share_recipients and threshold are internally coherent
+    let total_share_count = share_recipients
+        .iter()
+        .fold(0, |acc, (_, share_count)| acc + share_count);
+    if total_share_count < threshold {
+        return Err(CertifShamirError::InvalidThreshold);
+    }
+
+    if share_recipients
+        .keys()
+        .any(|&recipient| recipient == author_id)
+    {
+        return Err(CertifShamirError::AuthorIncludedAsRecipient);
+    }
+
+    if threshold == 0 {
+        return Err(CertifShamirError::ThresholdIsZero);
+    }
+
+    if share_recipients.values().any(|&v| v == 0) {
+        return Err(CertifShamirError::ShareRecipientHasZeroShares);
+    }
+
+    // 2. Check for previous setup
+    if let Some(setup) =
+        get_latest_shamir_setup_for_author(certificate_ops, &author_id, &timestamp).await?
+    {
+        return Err(CertifShamirError::ShamirSetupAlreadyExist(setup.timestamp));
+    }
+
+    // 3. Check organization's and author's status
+
+    // 4. Check recipients status
+
+    let mut participants_id: HashSet<_> = share_recipients.keys().collect();
+    participants_id.insert(&author_id);
+    // TODO implement a get_users method instead of list_user (let the DB do its job)
+    let participants = certificate_ops.list_users(true, None, None).await?;
+    let participants = participants
+        .iter()
+        .filter(|info| participants_id.contains(&info.id));
+
+    // no participant is missing
+    for &&id in &participants_id {
+        if !participants.clone().any(|info| info.id == id) {
+            return Err(CertifShamirError::MissingUser(id));
+        }
+    }
+
+    // all participants not revoked
+    for info in participants {
+        if info.revoked_on.is_some() {
+            return Err(CertifShamirError::UserRevoked(info.id));
+        }
+    }
+    // 5. Generate certificates
+
+    let brief = ShamirRecoveryBriefCertificate {
+        author: author_device,
+        timestamp,
+        user_id: author_id,
+        threshold: NonZeroU64::new(threshold.into()).expect("shamir threshold is zero"), // checked during the first step
+        per_recipient_shares: share_recipients
+            .iter()
+            .map(|(&k, &v)| (k, NonZeroU64::new(v.into()).expect("Share count is zero")))
+            .collect(),
+    }
+    .dump_and_sign(&certificate_ops.device.signing_key)
+    .into();
+
+    let recovery_device = LocalDevice::generate_new_device(
+        certificate_ops.device.organization_addr.clone(),
+        certificate_ops.device.initial_profile,
+        certificate_ops.device.human_handle.clone(),
+        DeviceLabel::default(),
+        Some(certificate_ops.device.user_id),
+        None,
+        None,
+        None,
+    );
```

Comment:
A new local device is generated but it doesn't seem to be registered on the server (using a `DeviceCertificate` and the `device_create` command)